### ISSUE TITLE
How to run MEND on custom data/ custom model?

### ISSUE THREAD

**Issue by sameeravithana** [eric-mitchell/mend#1]
Hello, just curious whether there are any documentation on how to use MEND framework on custom LM (e.g., gpt-2) trained with custom data?

**Comment by eric-mitchell** [comment]
Hi Sameera,

At the moment there's no documentation for this, but I'm happy to help get you up and running.

If your model's fine-tuned weights are in the HuggingFace Hub (see [this link](https://huggingface.co/docs/hub/main#how-can-i-loadpush-fromto-the-hub) for info on the hub and how to add models to it), all you should need to do to get the model working is write a new config under `config/model` with the name of the model checkpoint and tokenizer. If your model isn't in the HuggingFace Hub, there will be some code changes necessary in order to get training working, but mostly superficial things like adjusting tokenizers/loss functions/model loading in `trainer.py` and `model.py` to be compatible with your model's function signature and return types.

To use custom data, we'll need to implement a new data class in `data_classes` that loads your data and returns the tuples of data (edit input, edit label, outer input, outer label, locality example) needed for MEND training.

If you share some more details about your model's implementation and the data you'd like to train it with, we can figure out what the next step is.

**Comment by sameeravithana** [comment]
Hi Eric, We use gpt-neox to train gpt-2 style models with custom BPE tokenizers. I hope they would be very similar to other trained models available in HF. (we can not publish it in HF). We can prepare a qa dataset similar to nq. I can look into `trainer.py` and `model.py` given your suggestions for the modifications.

**Comment by eric-mitchell** [comment]
If the gpt-neox model just returns the predicted logits for each time step, it should be mostly plug and play with the current MEND implementation. However, the MEND code currently assumes [here](https://github.com/eric-mitchell/mend/blob/main/algs/mend.py#L181) that your model accepts inputs in the HuggingFace model format (`input_ids`, `attention_mask`). If this isn't the case, you might need to adjust that line. You can make your model editable as
    
    from algs.mend import MEND

    gpt_neox = load_your_model()  # whatever you normally do to create/init your model & weights
    editable_gpt = MEND(gpt_neox, config, lambda: copy.deepcopy(gpt_neox))

where `config` is a hydra config object and `config.mend` should exist and contain settings for the attributes listed in `config/alg/mend.yaml`. Call `editable_gpt.edit(batch)` to edit the model, where `batch` is a dict with `input_ids`, `attention_mask`, and `labels`, all PyTorch tensors. Before you can apply edits, though, you'll need to train the editable model using your data.

[Here](https://github.com/eric-mitchell/mend/blob/main/run.py#L68) is where we create the editable models for our experiments.

For data, you might be able to get by working from our zsre question answering examples (although these are entity-oriented). You'll just need to re-format the questions to be in the format for GPT, rather than BART/T5. Ultimately you'll need to implement a data class that implements an `edit_generator` like the one defined [here](https://github.com/eric-mitchell/mend/blob/main/data_classes/zsre.py#L112). The `edit_inner` and `edit_outer` tokens can technically be the same, but using different but related inputs/outputs in the outer loop can improve edit generalization.

If your dataset doesn't have paired questions that are rephrases of each other, you might benefit from mixing your QA dataset with the zsre dataset in `data_classes/zsre.py` so that you can take advantage of the question rephrases that zsre already has generated to improve the generalization performance of your edits.

Again, let me know if you have any questions about any of these steps- happy to help if I can!

**Comment by eric-mitchell** [comment]
Closing due to inactivity (but feel free to re-open if there are any new/lingering questions!).
